aten/src/ATen/core/dispatch/OperatorEntry.cpp

```diff
@@ -220,7 +220,15 @@ void OperatorEntry::updateDispatchTableEntry_(const c10::Dispatcher& dispatcher,
 }
 
 void OperatorEntry::updateDispatchTable_(const c10::Dispatcher& dispatcher, DispatchKey dispatch_key) {
-  for (auto k : c10::getRuntimeDispatchKeys(dispatch_key)) {
+  // Undefined isn't a runtime key but we have an entry for it in dispatchTable_ (which is not necessary).
+  // Update it separately if required.
+  // XXX: in fact I think we should just start from iter=1 in updateDispatchTableFull_.
+  // dispatchTable should never care about Undefined key.
+  if (dispatch_key == DispatchKey::Undefined) {
+    updateDispatchTableEntry_(dispatcher, dispatch_key);
+    return;
+  }
+  for (auto k : c10::getRuntimeDispatchKeySet(dispatch_key)) {
```

Comment:
Is getRuntimeDispatchKeySet a hot function? If it isn't we should make it raise an exception if it's passed Undefined.

Reply:
It's currently in hot path (that every dispatcher call go through it). We can remove it from hot path and only keeps it in precompute if we could move the autograd keyset constexpr to header file. But I'll need to run the ios binary size test on that so I'll do it in a followup PR. :D

Reply:
Wait, what? That shouldn't be true.

```
(/home/ezyang/local/pytorch-tmp-env) [ezyang@devvm066.ash0 ~/local/pytorch-tmp] git grep getRuntimeDispatchKeySet
aten/src/ATen/core/LegacyTypeDispatch.h:    autograd_guard_(getRuntimeDispatchKeySet(DispatchKey::Autograd)) {
aten/src/ATen/templates/TensorBody.h:  return c10::impl::tls_local_dispatch_key_set().excluded_.isSupersetOf(c10::getRuntimeDispatchKeySet(DispatchKey::Autograd));
c10/core/DispatchKeySet.cpp:DispatchKeySet getRuntimeDispatchKeySet(DispatchKey t) {
c10/core/DispatchKeySet.cpp:  return k != DispatchKey::Undefined && getRuntimeDispatchKeySet(alias).has(k);
c10/core/DispatchKeySet.h:C10_API DispatchKeySet getRuntimeDispatchKeySet(DispatchKey t);
c10/core/DispatchKeySet.h://      we can remove this API and use c10::getRuntimeDispatchKeySet instead.
c10/core/DispatchKeySet.h:// getRuntimeDispatchKeySet(alias).has(DispatchKey::Undefind)
c10/core/DispatchKeySet.h:  return (s - getRuntimeDispatchKeySet(DispatchKey::Autograd)).highestPriorityTypeId();
c10/core/TensorImpl.h:    key_set_ = key_set_ - getRuntimeDispatchKeySet(DispatchKey::Autograd);
c10/core/impl/LocalDispatchKeySet.cpp:      raw_local_dispatch_key_set.excluded() | getRuntimeDispatchKeySet(DispatchKey::Autograd));
```

I don't think any of these are on dispatcher calls? (It looks like it does show up in autoNonVariableTypeMode). The autograd set should probably get constexpr'ed

Reply:
Yup `AutoNonVariableType` is the one I had in mind. Btw I updated the PR to separate autograd set to be a separate function returning a cheap copy of our constexpr (still keep the constexpr variable in cpp file so that we don't increase ios binary size). Please let me know if there's a better way to do it :D  Thanks!

Reply:
I'm still confused. DispatchKeySet is one word. If it's all constexpr there shouldn't be any binary size problem putting it in a header. Am I missing something?

Reply:
cc @bhosmer